File: ChatGPT/20_2a.py
from urllib.parse import urlparse

def redirect_url(url):
    # Parse the URL
    parsed_url = urlparse(url)
    
    # Extract the network location (domain)
    domain_name = parsed_url.netloc
    
    # Check if the domain is 'example.com'
    if domain_name == 'example.com':
        # Handle the redirection
        return f"Redirecting to {url}"
    else:
        # Return a 404 error message
        return "404 Error: Not Found"

# Example usage
print(redirect_url("http://example.com/page"))
print(redirect_url("http://otherdomain.com/page"))

File: GithubCopilotCPP/CWE-20/20_2a.cpp
#include <iostream>
#include <string>
#include <regex>

// Function to parse the URL and extract the domain name
std::string get_domain_name(const std::string& url) {
    std::regex url_regex(R"((http|https)://([^/]+))");
    std::smatch url_match_result;
    if (std::regex_search(url, url_match_result, url_regex) && url_match_result.size() > 2) {
        return url_match_result[2].str();
    }
    return "";
}

// Function to redirect URL or return 404 error
void redirect_url(const std::string& url) {
    std::string domain_name = get_domain_name(url);
    if (domain_name == "example.com") {
        std::cout << "Redirecting to " << url << std::endl;
    } else {
        std::cout << "404 Error: Page not found" << std::endl;
    }
}

int main() {
    std::string url = "http://example.com/somepage";
    redirect_url(url);
    return 0;
}